flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/sort/SortingDataInput.java

```diff
@@ -0,0 +1,227 @@
+ * from the chained input. Moreover the timestamps of incoming records are used for secondary ordering.
+ * For the comparison it uses either {@link FixedLengthByteKeyComparator} if the length of the
+ * serialized key is constant, or {@link VariableLengthByteKeyComparator} otherwise.
+ *
+ * <p>Watermarks, stream statuses, nor latency markers are not propagated downstream as they do not make
+ * sense with buffered records. The input emits a MAX_WATERMARK after all records.
+ *
+ * @param <T> The type of the value in incoming {@link StreamRecord StreamRecords}.
+ * @param <K> The type of the key.
+ */
+public final class SortingDataInput<T, K> implements StreamTaskInput<T> {
+
+	private final StreamTaskInput<T> chained;
+	private final PushSorter<Tuple2<byte[], StreamRecord<T>>> sorter;
+	private final KeySelector<T, K> keySelector;
+	private final TypeSerializer<K> keySerializer;
+	private final DataOutputSerializer dataOutputSerializer;
+	private final ForwardingDataOutput forwardingDataOutput;
+	private MutableObjectIterator<Tuple2<byte[], StreamRecord<T>>> sortedInput = null;
+	private boolean emittedLast;
+
+	public SortingDataInput(
+			StreamTaskInput<T> chained,
+			TypeSerializer<T> typeSerializer,
+			TypeSerializer<K> keySerializer,
+			KeySelector<T, K> keySelector,
+			MemoryManager memoryManager,
+			IOManager ioManager,
+			boolean objectReuse,
+			double managedMemoryFraction,
+			Configuration jobConfiguration,
+			AbstractInvokable containingTask) {
+		try {
+			this.forwardingDataOutput = new ForwardingDataOutput();
+			this.keySelector = keySelector;
+			this.keySerializer = keySerializer;
+			int keyLength = keySerializer.getLength();
+			final TypeComparator<Tuple2<byte[], StreamRecord<T>>> comparator;
+			if (keyLength > 0) {
+				this.dataOutputSerializer = new DataOutputSerializer(keyLength);
+				comparator = new FixedLengthByteKeyComparator<>(keyLength);
+			} else {
+				this.dataOutputSerializer = new DataOutputSerializer(64);
+				comparator = new VariableLengthByteKeyComparator<>();
+			}
+			KeyAndValueSerializer<T> keyAndValueSerializer = new KeyAndValueSerializer<>(typeSerializer, keyLength);
+			this.chained = chained;
+			this.sorter = ExternalSorter.newBuilder(
+					memoryManager,
+					containingTask,
+					keyAndValueSerializer,
+					comparator)
+				.memoryFraction(managedMemoryFraction)
+				.enableSpilling(
+					ioManager,
+					jobConfiguration.get(AlgorithmOptions.SORT_SPILLING_THRESHOLD))
+				.maxNumFileHandles(jobConfiguration.get(AlgorithmOptions.SPILLING_MAX_FAN))
+				.objectReuse(objectReuse)
+				.largeRecords(true)
+				.build();
+		} catch (MemoryAllocationException e) {
+			throw new RuntimeException(e);
+		}
+	}
+
+	@Override
+	public int getInputIndex() {
+		return chained.getInputIndex();
+	}
+
+	@Override
+	public CompletableFuture<Void> prepareSnapshot(
+			ChannelStateWriter channelStateWriter,
+			long checkpointId) throws IOException {
```

Comment:
My IDE is warning me about a bunch of exceptions that are declared but never thrown in these methods.